src/gs-category-page.c

```diff
@@ -230,6 +235,97 @@ copy_os_to_usb_button_cb (GtkButton *button, GsCategoryPage *self)
 			 self->cancellable);
 }
 
+static gchar*
+removable_destination_dup (GsCategoryPage *self)
+{
+	/* first one is the most-recently added mount on a removable disk */
+	if (self->copy_dests)
+		return g_strdup (self->copy_dests->data);
+
+	return NULL;
+}
+
+static void
+gs_category_page_os_get_copyable_cb (GObject *source_object,
+				     GAsyncResult *res,
+				     gpointer user_data)
+{
+	GsPluginLoader *plugin_loader = GS_PLUGIN_LOADER (source_object);
+	GsCategoryPage *self = GS_CATEGORY_PAGE (user_data);
+	g_autoptr(GError) error = NULL;
+	gboolean copyable;
+
+	copyable = gs_plugin_loader_job_os_get_copyable_finish (plugin_loader,
+								res, &error);
+	gtk_widget_set_visible (self->usb_action_box, copyable);
+}
+
+static void
+gs_category_page_update_os_copy_button (GsCategoryPage *self)
+{
+	g_autofree gchar* copy_dest = NULL;
+	g_autoptr(GsPluginJob) plugin_job = NULL;
+
+	if (self->category == NULL ||
+	    g_strcmp0 (gs_category_get_id (self->category), "usb") != 0) {
+		gtk_widget_set_visible (self->usb_action_box, FALSE);
+		return;
+	}
+
+	gtk_widget_set_visible (self->usb_action_box, TRUE);
+	gtk_widget_set_sensitive (self->copy_os_to_usb_button, FALSE);
+	gtk_widget_set_visible (self->copy_os_to_usb_button, FALSE);
+
+	if (self->copying)
+		return;
+
+	copy_dest = removable_destination_dup (self);
+	if (copy_dest != NULL) {
+		gtk_button_set_label (GTK_BUTTON (self->copy_os_to_usb_button), _("Copy OS to USB"));
+		gtk_widget_set_sensitive (self->copy_os_to_usb_button, TRUE);
+
+		/* also check (asynchronously) whether the OS will
+		 * certainly fail a copy so we can update the UI
+		 * accordingly. Cases where the plugin would know a copy
+		 * would fail include an OSTree remote which lacks a collection
+		 * ID or broader issues like the user lacking write
+		 * access to the destination. */
```

Comment:
I don’t see any code which checks whether the user has write access to the destination.

Reply:
Would you really want to hide the button in that case? It seems more user friendly to show the button, and then present the user with a suitable error message. Or, even better, chmod the destination so we actually do have write access.

Conversely, I think it makes sense to disable the button if no collection ID is set, because that’s beyond the control of any (non-technical) user.

Reply:
To be consistent with the behavior when no collection ID is configured on an app remote, the button shouldn't show at all.

Reply:
> I don’t see any code which checks whether the user has write access to the destination.

None exists; that's just an example of what could be done. It's mostly highlighting the design of the API in case other plugin developers want to support that use case (or in case we eventually do).

> Would you really want to hide the button in that case? It seems more user friendly to show the button, and then present the user with a suitable error message. Or, even better, chmod the destination so we actually do have write access.

I can see that argument though what I currently do follows the convention used in similar contexts (eg, app copying which follows the convention for making an app's desktop shortcut, etc.).

The UX needs review from a designer in general, so I'd prefer to follow convention and punt until we get to that point.

As far as ensuring the proper permissions, as of recently, we ensure the `.ostree` dir is writable whenever we make it. If the `.ostree` dir doesn't exist and we don't have write access, I don't think we should try `chmod`ing the parent dir.

> Conversely, I think it makes sense to disable the button if no collection ID is set, because that’s beyond the control of any (non-technical) user.

I agree if we're showing the button as you suggested above but let's address that if/when we reverse the convention.